File: combinations.py
'''
Time: O(N^(T/M + 1)) where T is the target, M is the minimum element in the candidates
Space: O(T/M) or O(T)
'''
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        global output
        output=[]
        path=[]
        
        if candidates is None or len(candidates)==0:
            return output
        
        self.backtracking(candidates,target,path,0)
        return output
    
    def backtracking(self,candidates,target,path,index):
        if target<0:
            return 
        
        if target==0:
            output.append(list(path)) #Deep Copy 
            return
        
        #Backtracking
        for i in range(index,len(candidates)):
            path.append(candidates[i])
            self.backtracking(candidates,target-candidates[i],path,i)
            path.pop()
        

File: combinations.java
/*
Time complexity: O(2^mn)
Space complexity: O(mn)
*/
import java.util.*;
class Solution {
    void helper(int[] array,int sum,int start, List<Integer> path, List<List<Integer>> res){
        //base condition
        if(sum<0){
            return ;
        }
        if(sum==0){
            res.add(new ArrayList<>(path));
            return;
        }
        
        //branching condition.
        for(int i=start;i<array.length;i++){
            path.add(array[i]);
            helper(array,sum-array[i],i,path,res);
             path.remove(path.size()-1);       
        }
       
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
    List<List<Integer>> res=new ArrayList<List<Integer>>();
    List<Integer> path=new ArrayList<Integer>();
    helper(candidates,target,0,path,res);
    return res;
    }
}